### PR TITLE
setup: detailed error messages

### PR DESCRIPTION
**Pull request by incognitojam**
**Generic download failure:**
Screenshot from 2023-02-23 17-25-04

**Download a non-executable file:**
Screenshot from 2023-02-23 17-25-20

### CODE REVIEW

File: selfdrive/ui/qt/setup/setup.cc [hunk 1]
```diff
@@ -309,19 +368,42 @@ Setup::Setup(QWidget *parent) : QStackedWidget(parent) {
...
+        break;
+      case DownloadResult::error:
+      default:
+        setCurrentWidget(error_widget);
+        break;
     }
   });
 
+  // load fonts
+  QFontDatabase::addApplicationFont("../assets/fonts/Inter-Black.ttf");
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/setup/setup.cc:397]: this won't work in setup. are you sure this is necessary on C3? might just be your workstation

File: selfdrive/ui/qt/setup/setup.cc [hunk 2]
```diff
@@ -276,12 +280,67 @@ QWidget * Setup::download_failed() {
...
+  main_layout->addSpacing(67);
+
+  url->setWordWrap(true);
+  url->setAlignment(Qt::AlignTop | Qt::AlignLeft);
+  url->setStyleSheet("font-family: \"JetBrains Mono\"; font-size: 64px; font-weight: 400; margin-right: 100px;");
+  main_layout->addWidget(url);
+
+  main_layout->addSpacing(60);
+
+  QLabel *body = new QLabel(tr("Ensure the entered URL is valid."));
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/setup/setup.cc:313]:
```suggestion
  QLabel *body = new QLabel(tr("No custom software found at this URL."));
```

File: selfdrive/ui/qt/setup/setup.cc [hunk 3]
```diff
@@ -276,12 +280,67 @@ QWidget * Setup::download_failed() {
...
+
+  QLabel *title = new QLabel(tr("Invalid Custom Software URL"));
+  title->setStyleSheet("font-size: 90px; font-weight: 500;");
+  main_layout->addWidget(title, 0, Qt::AlignTop | Qt::AlignLeft);
+
+  main_layout->addSpacing(67);
+
+  url->setWordWrap(true);
+  url->setAlignment(Qt::AlignTop | Qt::AlignLeft);
+  url->setStyleSheet("font-family: \"JetBrains Mono\"; font-size: 64px; font-weight: 400; margin-right: 100px;");
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/setup/setup.cc:308]: oh this probably isn't in agnos. you can just open an agnos-builder PR to add

File: selfdrive/ui/qt/setup/setup.cc (hunk 1 above)
Comment (incognitojam) [reply selfdrive/ui/qt/setup/setup.cc:397]: I did get it to work on C3 and checked the QFontDatabase only listed `JetBrains Mono` after importing it this way, but you are right we should add it to agnos

**Comment by incognitojam** [comment]
related: https://github.com/commaai/agnos-builder/pull/113

File: selfdrive/ui/qt/setup/setup.cc [hunk 4]
```diff
@@ -287,6 +290,61 @@ QWidget * Setup::download_failed() {
   return widget;
 }
 
+QWidget * Setup::download_invalid_url(QLabel *url) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/setup/setup.cc:293]: can't we reuse the failed widget for this? think it makes sense to show the URL there too anyway

File: selfdrive/ui/qt/setup/setup.cc [hunk 5]
```diff
@@ -312,13 +370,25 @@ Setup::Setup(QWidget *parent) : QStackedWidget(parent) {
...
-      QTimer::singleShot(3000, this, &QWidget::hide);
-    } else {
-      setCurrentWidget(failed_widget);
+  QLabel *url_label = new QLabel();
+  invalid_url_widget = download_invalid_url(url_label);
+  addWidget(invalid_url_widget);
+
+  QObject::connect(this, &Setup::finished, [=](const DownloadResult &result, const QString &url) {
+    qDebug() << "finished" << result << url;
+    switch (result) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/setup/setup.cc:379]: using the failed widget also gets rid of this switch

**Review by adeebshihadeh — APPROVED** [review]
Looks a lot cleaner!

**Pull request merged**